// @nktkas/hyperliquid@0.26.0/dist/src/api/subscription/spotState.js
import * as v from "valibot";
import { Address, parser } from "../_base.js";
import { BalanceSchema } from "../_common_schemas.js";
// -------------------- Schemas --------------------
/** Subscription to spot state events for a specific user. */
export const SpotStateRequest = /* @__PURE__ */ (() => {
    return v.pipe(v.object({
        /** Type of subscription. */
        type: v.pipe(v.literal("spotState"), v.description("Type of subscription.")),
        /** User address. */
        user: v.pipe(Address, v.description("User address.")),
    }), v.description("Subscription to spot state events for a specific user."));
})();
/** Event of user spot state. */
export const SpotStateEvent = /* @__PURE__ */ (() => {
    return v.pipe(v.object({
        /** User address. */
        user: v.pipe(Address, v.description("User address.")),
        /** Spot state of the user. */
        spotState: v.pipe(v.object({
            /** Balance for each token. */
            balances: v.pipe(v.array(BalanceSchema), v.description("Balance for each token.")),
        }), v.description("Spot state of the user.")),
    }), v.description("Event of user spot state."));
})();
/**
 * Subscribe to spot state updates for a specific user.
 * @param config - General configuration for Subscription API subscriptions.
 * @param params - Parameters specific to the API subscription.
 * @param listener - A callback function to be called when the event is received.
 * @returns A request-promise that resolves with a {@link Subscription} object to manage the subscription lifecycle.
 *
 * @throws {TransportError} When the transport layer throws an error.
 *
 * @see https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/websocket/subscriptions
 * @example
 * ```ts
 * import { WebSocketTransport } from "@nktkas/hyperliquid";
 * import { spotState } from "@nktkas/hyperliquid/api/subscription";
 *
 * const transport = new WebSocketTransport();
 *
 * const sub = await spotState(
 *   { transport },
 *   { user: "0x..." },
 *   (data) => console.log(data),
 * );
 * ```
 */
export function spotState(config, params, listener) {
    const payload = parser(SpotStateRequest)({ type: "spotState", user: params.user });
    return config.transport.subscribe(payload.type, payload, (e) => {
        if (e.detail.user === payload.user) {
            listener(e.detail);
        }
    });
}
